keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,868 @@
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP aims to add a new field named `maxRestartTimes`(TBD) to podSpec which helps controlling
+the restart times of a container when pod's `restartPolicy` is set to `OnFailure`.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+Currently, we have backoffLimit to fail a Job after some amount of retries, which helps to
+prevent unnecessary restarts. `Pod` is also eager for this feature to avoid infinite restarting
+due to a system internal error etc. What's more, considering `Job` is working at a higher level,
+adding `maxRestartTimes` to `Pod` might be reusable.
+
+As a matter of fact, it's a long-standing issue and strongly needed by the community, see [issue](https://github.com/kubernetes/kubernetes/issues/65797).
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Add a new field `maxRestartTimes` to the spec of pod which only applies when `restartPolicy` is `OnFailure`
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Combine the implementations of `backoffLimit` and `maxRestartTimes` together.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new field `maxRestartTimes` to the spec of pod. `maxRestartTimes` only applies
+when `restartPolicy` is set to `OnFailure`. As a result, every container of
+the pod has the limit restart times defined by `maxRestartTimes`.
```

Comment:
Should it be the limit times for each container, or the limit of their total restart times? The later one might be more understandable if the field is added into PodSpec?

Reply:
Good catch, I think you're right, the original `RestartCount` is the total restart times of the containers, so the `maxRestartTimes` should also respect all containers. Thanks.

Reply:
Good catch. If we found a container reach the limit restart times, and we think it's unrecoverable, restart the other containers makes no sense. Thanks.

Reply:
Updated.